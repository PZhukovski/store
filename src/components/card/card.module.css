.block {
  display: flex;
  padding-top: 30px;
  gap: 30px;
  margin-bottom: 100px;
}
.slider_group{
  width: 460px;
  margin-left: 80px;
}
.wrapper_image {
  position: relative;
  width: 100%;
  height: 100%;
  margin-bottom: 133%;
}
.image {
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  width: 460px;
  height: 100%;
  transition: background-image 0.2s
}
.images_gallery {
  display: flex;
}
.image_small {
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
  width: 56px;
  height: 56px;
  /* padding-bottom: 100%; */
  transition: opacity ease-in-out 0.3s;
  margin-right: 5px;
  margin-top: 4px;
}
.select {
  min-width: 150px;
  position: relative;
  display: table;
}
.description {
  max-width: 460px;
}
@media (min-width: 1240px) {
  .slider_group{
    margin-left: 180px;
  }
}
@media (max-width: 1200px) {
  .description {
    max-width: 380px;
  }
}
@media (max-width: 950px) {
  .block {
    display: flex;
    flex-direction: column;
    padding-top: 30px;
    gap: 30px;
    margin: 0 auto;
  }
  .wrapper_image {
    margin-bottom: 0%;
  }
  .slider_group{
    max-width: 600px;
    margin: 0;
    max-width: none;
    width: auto;
    /* margin-left: 80px; */
    /* margin: 0 auto; */
  }
  .image {
    width: 100%;
    background-size: contain;
    height: 0;
    padding-top: 120%;
  }
  .images_gallery {
    justify-content: center;
  }
  .description {
    margin: 0 auto;
    max-width: 540px;
  }
}
