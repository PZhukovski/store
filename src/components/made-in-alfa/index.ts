export * from "./made-in-alfa"