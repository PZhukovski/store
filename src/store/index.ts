import {configureStore} from '@reduxjs/toolkit' ;
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import madeInAlfa from '../slices/made-in-alfa-slice'
import  yourDesign  from '../slices/your-design-slice';
import  product  from '../slices/product-slice';
import  cart  from '../slices/cart-slice';
import {ThunkDispatch} from "@reduxjs/toolkit";

export const store = configureStore({
    reducer: {
      madeInAlfa, yourDesign, product, cart
    },
    middleware: (getDefaultMiddleware: () => any) => getDefaultMiddleware(),
    devTools: process.env.NODE_ENV !== 'production',

})

export type ApplicationState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<ThunkDispatch<any, any, any>>();
// export const useAppDispatch = () => useDispatch<typeof store.dispatch>()
export const useAppSelector: TypedUseSelectorHook<ApplicationState> = useSelector;
