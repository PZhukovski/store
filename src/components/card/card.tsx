import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { fetchProduct, product } from "../../slices/product-slice";
import { Typography } from "@alfalab/core-components/typography";
import { Gap } from "@alfalab/core-components/gap";
import { Button } from "@alfalab/core-components/button";
import { SelectLayout } from "../select-layout";
import { useTypeSelector } from "../../hooks/use-type-selector";
import { addToCart } from "../../slices/cart-slice";
import { useForm, SubmitHandler } from "react-hook-form";
import { v1 } from 'uuid';
import { useAppDispatch } from "../../store";
// @ts-ignore
import styles from "./card.module.css";

type HooksPropsType = {
    size:  string | undefined;
    color:  string | undefined;
    sticker: number | undefined;
};
export const Card = () => {

  const {id} = useParams();
  const idMemo = Number(id)
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchProduct(idMemo))
  }, [idMemo]);

  const item = useTypeSelector(product)

  useEffect(() => {
   setActiveSlide(preview)

  }, [item]);

  const {
    preview,
    images,
    title,
    stickerNumbers,
    description,
    price,
    sizes,
    colors,
  } = item[0] || {};

  const [activeSlide, setActiveSlide] = useState(preview);

  const initialState : HooksPropsType = {
    size:  undefined,
    color:  undefined,
    sticker:  undefined
  };
  const { handleSubmit, control , reset} = useForm({defaultValues: initialState});
  const handleImageChange = (image: string) => () => {
    setActiveSlide(image);
  };

  const handleAddBusket : SubmitHandler<any>= (data) => {
    const {size, color, sticker} = data;
    const itemToBucket = {
      id: Number(id),
      title: title,
      color: color!==undefined ? color.selected.content : null,
      size: size!==undefined ? size.selected.content : null,
      stickerNumber: sticker!==undefined ? sticker.selected.content : null,
      preview: preview,
      price: price,
      quantity: 0,
      sum: price,
      removeId: v1()
    };
    dispatch(addToCart(itemToBucket))
    reset(initialState)
  };

  return (
    <>
      <div className={styles.block}>
        <div className={styles.slider_group}>
          <div className={styles.wrapper_image}>
            <div
              className={styles.image}
              style={{ backgroundImage: `url(${activeSlide})` }}></div>
            <div className={styles.images_gallery}>
              {images?.map((image, index) => {
                return (
                  <div
                    key={index}
                    onClick={handleImageChange(image)}
                    className={styles.image_small}
                    style={{ backgroundImage: `url(${image})` }}></div>
                );
              })}
            </div>
          </div>
        </div>
        <form onSubmit = {handleSubmit(handleAddBusket)}>
        <div className={styles.description}>
          <Typography.TitleResponsive view="small" font="styrene" tag="h4">
            {title}
          </Typography.TitleResponsive>
          <Gap size="m" />
          <Typography.TitleResponsive
            view="small"
            font="styrene"
            weight="bold"
            tag="h4">
            {price} ₽
          </Typography.TitleResponsive>
          <SelectLayout
            title={"размер"}
            name ={"size"}
            entity={sizes}
            control ={control}
          />
          <SelectLayout
            title={"цвет"}
            name ={"color"}
            entity={colors}
            control ={control}
          />
          <SelectLayout
            title={"номер стикера"}
            name ={"sticker"}
            entity={stickerNumbers}
            control ={control}
          />
          <Gap size="m" />
          <Button
            type ="submit"
            view="primary"
            size="xs">
            В корзину
          </Button>
          <Gap size="m" />
          <Typography.Text tag="div" weight="medium" view="secondary-medium">
            {description}
          </Typography.Text>
        </div>
        </form>
      </div>
    </>
  );
};


