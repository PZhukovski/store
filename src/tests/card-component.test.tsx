import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Card } from '../components/card';
import { Provider } from "react-redux";
import { store } from "../store/index";

test('snapshot Card', async() => {
  const cardElement = await render(<Provider store={store}><Card/></Provider>);
  expect(cardElement).toMatchSnapshot();
});
