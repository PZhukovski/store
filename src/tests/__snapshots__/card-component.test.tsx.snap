// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Card 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="block"
      >
        <div
          class="slider_group"
        >
          <div
            class="wrapper_image"
          >
            <div
              class="image"
              style="background-image: url(undefined);"
            />
            <div
              class="images_gallery"
            />
          </div>
        </div>
        <form>
          <div
            class="description"
          >
            <h4
              class="typography__component_mub95 typography__styrene-small_g9g7d typography__medium_mub95"
            />
            <div
              class="gap__gap_15oeq gap__vertical_15oeq"
              data-gap-size="m"
            />
            <h4
              class="typography__component_mub95 typography__styrene-small_g9g7d typography__bold_mub95"
            >
               ₽
            </h4>
            <div
              class="gap__gap_15oeq gap__vertical_15oeq"
              data-gap-size="m"
            />
            <button
              class="button__component_s1pl2 button__xs_s1pl2 button__component_17krt button__primary_17krt"
              type="submit"
            >
              <span
                class="button__text_s1pl2 button__stretchText_s1pl2"
              >
                В корзину
              </span>
            </button>
            <div
              class="gap__gap_15oeq gap__vertical_15oeq"
              data-gap-size="m"
            />
            <div
              class="typography__secondary-medium_11grk typography__medium_11grk"
            />
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="block"
    >
      <div
        class="slider_group"
      >
        <div
          class="wrapper_image"
        >
          <div
            class="image"
            style="background-image: url(undefined);"
          />
          <div
            class="images_gallery"
          />
        </div>
      </div>
      <form>
        <div
          class="description"
        >
          <h4
            class="typography__component_mub95 typography__styrene-small_g9g7d typography__medium_mub95"
          />
          <div
            class="gap__gap_15oeq gap__vertical_15oeq"
            data-gap-size="m"
          />
          <h4
            class="typography__component_mub95 typography__styrene-small_g9g7d typography__bold_mub95"
          >
             ₽
          </h4>
          <div
            class="gap__gap_15oeq gap__vertical_15oeq"
            data-gap-size="m"
          />
          <button
            class="button__component_s1pl2 button__xs_s1pl2 button__component_17krt button__primary_17krt"
            type="submit"
          >
            <span
              class="button__text_s1pl2 button__stretchText_s1pl2"
            >
              В корзину
            </span>
          </button>
          <div
            class="gap__gap_15oeq gap__vertical_15oeq"
            data-gap-size="m"
          />
          <div
            class="typography__secondary-medium_11grk typography__medium_11grk"
          />
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
